⚙️ Eslint & Husky+lint-staged 설정

Legitgoons opened the pull request:
## 작업 이유
- Husky와 lint-staged를 사용한 pre-commit 단계에서의 lint 실행 구현
  - 현재는 CI에서 eslint를 catch하는 방식
  - CI과정에서 모든 코드를 한 번에 확인하고 수정하는 것은 많은 시간과 리소스 소모
- 합의된 Eslint 규칙 추가
<br/>

## 작업 사항
### 1️⃣ Husky와 lint-staged를 사용한 pre-commit 단계에서의 lint 실행
- .husky/pre-commit을 통해 commit 전에 문구가 출력되고, lint-front가 실행되도록 설정
```
//.husky/pre-commit
echo "🐶 Running lint-staged..."
yarn lint-front
```
- package.json에서는 lint-front가 실행되면, lint-staged가 동작하도록 구현
- lint-staged에서는 eslint --fix와 prettier --write를 실행
```json
"scripts": {
  "postinstall": "husky",
  "lint-front": "lint-staged"
},
"lint-staged": {
  "*.{js,jsx,ts,tsx}": "eslint --fix",
  "*.{js,jsx,ts,tsx,css,md}": "prettier --write"
}
```
- commit시 자동 실행
![image](https://github.com/CollaBu/pennyway-client-webview/assets/101088491/d6966a5f-3d22-4f31-ba61-c0505fe4cca7)

### 2️⃣ 합의된 Eslint 규칙 추가
```cjs
"no-var": "error", // var 금지
"no-multiple-empty-lines": "error", // 여러 줄 공백 금지
"no-console": ["error", { "allow": ["warn", "error", "info"] }], // console.log() 금지
"eqeqeq": "error", // 일치 연산자 사용 필수
"dot-notation": "error", // 가능하다면 dot notation 사용
"no-unused-vars": "error" // 사용하지 않는 변수 금지
"react/jsx-no-useless-fragment": "error" // 불필요한 <React.Fragment>, <> 사용 금지
"react/react-in-jsx-scope": "off", // import React 입력 필요 X
```
#### import/order
```cjs
"groups": ["builtin", "external", "parent", "sibling", "index"],
// node.js의 내장모듈, 외부 설치 모듈, 상위 디렉토리 모듈, 현재/하위 디렉토리 모듈, 현재 디렉토리 인덱스 모듈
// 위의 순서대로 그룹화하여 정렬됩니다.
//...
'newlines-between': 'always', // 그룹 사이에 줄바꿈
alphabetize: {
  order: 'asc', // a~z순으로
  caseInsensitive: true, // 대소문자 구분 X
```
#### fsd-import
- fsd 아키텍처를 원활하게 사용하기 위해 설치했습니다.
```cjs
'fsd-import/fsd-relative-path': 'error',
'fsd-import/public-api-imports': 'error',
'fsd-import/layer-imports': 'error',
```
#### .eslintignore에 shared/mocks/* 추가
- Mock API의 경우 임시적으로 서버의 역할을 대체하는 역할입니다.
- 프로덕트처럼 엄격한 eslint rule을 적용할 필요가 없기에, eslintignore에 shared/mocks/*을 추가하였습니다.

### 3️⃣ 기타
#### components -> features 이름 변경
- 합의에 따라 components 폴더를 features로 이름을 변경하였습니다.
- 이에 따라 README의 이미지와 설명도 수정했습니다.
<br/>

## 리뷰어가 중점적으로 확인해야 하는 부분
- [x] Husky와 lint-staged가 정상적으로 작동하는지
- [x] 추가된 Eslint 규칙이 합의와 일치하는지
- [x] broswer.ts 파일에서 msw & import/order 이슈가 있다는 것을 인지했는지
- [x] components 폴더를 features로 이름을 변경한 것을 인지했는지

<br/>

## 발견한 이슈
- eslint의 rule Definition를 찾지 못하는 이슈
  - eslint-plugin-react를 설치하고, eslint의 plugins에 'react'를 추가해 해결
- eslint의 import/order에서 절대경로를 찾지 못하는 이슈
  - eslint-import-resolver-alias를 설치해 해결
- mocks/browser.ts에서 import/order rule과 충돌하는 코드가 있어 에러가 발생함
  - broswer.ts에서 import/order 활성화 시 msw의 경로로 인해 에러가 발생
  - eslint-import-resolver-alias가 원인으로 추측(참고 링크 : https://github.com/mswjs/msw/discussions/2062)
  - eslintignore에 shared/mocks/* 를 추가하는 것으로 해결

Legitgoons committed:
feat: husky, lint-staged 설치

Legitgoons committed:
feat: husky, lint-staged 동작 설정

package.json과 pre-commit 파일을 사용해 commit 전 lint를 실행하도록 설정하였습니다.

Legitgoons committed:
feat:  eslint-plugin-fsd-import 설치 및 적용

Legitgoons committed:
feat:  import 정렬을 위한 eslint-plugin-import 설치 및 적용

Legitgoons committed:
feat: 합의한 eslint rules 추가

Legitgoons committed:
fix: 이슈 해결을 위해 plugins 설치 및 eslint 설정

eslint의 rule Definition를 찾지 못하는 이슈가 있었습니다.
해결을 위해 eslint-plugin-react를 설치하고, eslint의 plugins에 'react'를 추가하였습니다.

Legitgoons committed:
fix: eslint의 import/order에서 절대경로를 찾지 못하는 이슈 해결을 위해 eslint-import-resolver-alias 설치

Legitgoons committed:
fix: eslint의 import/order rule에 알맞게 import 순서 수정

Legitgoons committed:
fix: browser.ts에서 에러 발생, import/order 임시 비활성화

broswer.ts에서 import/order 활성화 시 msw의 경로로 인해 에러가 발생합니다.
해당 이슈는 eslint-import-resolver-alias가 원인으로 추측됩니다.
우선 해당 파일에서 import/order를 비활성화하였습니다.

BangDori commented on .eslintrc.cjs:
P2: `fsd-import` 플러그인에 대한 `rules`가 혹시 정상적으로 적용이 되고 있나요?

Legitgoons replied:
앗 이걸 빠트렸네요. 추가하겠습니다!

BangDori commented on src/shared/mocks/browser.ts:
P3: `fsd-import`, `import/order` 등의 빡빡한 규칙을 두는 이유는 저희가 체계적으로 프로젝트를 관리하기 위함이라고 생각합니다.

하지만 Mock API의 경우 임시적으로 서버의 역할을 해주는 것이지, 클라이언트측에서 빡빡하게 관리하여야 할 대상이라고 생각되지는 않습니다. (실제로 현재 msw는 `development` 모드에서만 실행됩니다!)

그래서 Mocks와 관련한 파일들에 ESLint 규칙 적용을 제외하는 방향으로 가는 것도 좋을 것 같다고 생각하는데, 어떻게 생각하시나요?

Legitgoons replied:
동의합니다! Mocks 관련 파일들에서는 ESLint 규칙 적용을 제외하도록 하겠습니다.

Legitgoons replied:
shared/mocks의 파일들을 .eslintignore에 추가하고, mocks/browser.ts의 eslint-disable import/order 주석을 삭제했습니다!

BangDori reviewed (changes requested):
`fsd-import` 플러그인에 대한 규칙이 빠져있는 것 같아서, Request Changes 하였습니다!

확인 부탁드립니다~

Legitgoons committed:
feat: 합의에 따라 components 폴더를 features로 변경

Legitgoons committed:
docs: README fsd 아키텍처 이미지 및 설명 수정

Legitgoons committed:
feat: eslint rules에 fsd-import 추가

Legitgoons committed:
feat: shared/mocks의 파일들을 .eslintignore에 추가

합의에 따라 mocks 관련 파일들에 ESLint 규칙 적용을 제외하기 위해 .eslintignore에 추가했습니다.
또한 mocks/browser.ts의 eslint-disable import/order 주석도 삭제처리했습니다.

BangDori commented on .eslintignore:
P5: .eslintrc.cjs 내부에 ignorePatterns에 작성하는 건 어떤가요? lint를 제외하는 파일의 수가 많지않아서, eslintrc.cjs에서 관리해도 좋을 거 같아요!

Legitgoons replied:
.eslintrc.cjs에는 여러 관심사가 있기에, 여기서 관리하기보다는 eslintignore에서 바로 볼 수 있게 하는게 좀 더 직관적이라고 생각했습니다. 이 부분은 따로 수정하지 않겠습니당!

BangDori reviewed (approved):
음 아래 의견은 제 개인적인 의견이니, 반영하셔도 좋고 바로 Merge 하셔도 좋을 거 같습니다~

고생하셨습니다~
